lms/static/scripts/frontend_apps/utils/url.ts

```diff
@@ -0,0 +1,40 @@
+/**
+ * Polyfill for `Object.hasOwn`.
+ *
+ * `hasOwn(someObject, property)` should be used instead of
+ * `someObject.hasOwnProperty(name)`.
+ */
+function hasOwn(object: object, property: string): boolean {
+  return Object.prototype.hasOwnProperty.call(object, property);
+}
+
+/**
+ * Replace parameters in a URL template with values from a `params` object.
+ *
+ * Returns an object containing the expanded URL and a dictionary of unused
+ * parameters.
+ *
+ *   replaceURLParams('/things/:id', {id: 'foo', q: 'bar'}) =>
+ *     {url: '/things/foo', unusedParams: {q: 'bar'}}
+ */
+export function replaceURLParams<Param>(
+  urlTemplate: string,
+  params: Record<string, Param>,
+): { url: string; unusedParams: Record<string, Param> } {
+  const unusedParams: Record<string, Param> = {};
+  for (const param in params) {
+    if (hasOwn(params, param)) {
+      const value = params[param];
+      const urlParam = `:${param}`;
+      if (urlTemplate.indexOf(urlParam) !== -1) {
+        urlTemplate = urlTemplate.replace(
+          urlParam,
+          encodeURIComponent(String(value)),
+        );
+      } else {
+        unusedParams[param] = value;
+      }
+    }
+  }
```

Comment:
I note this copied function assumes there is exactly one usage of each parameter, and doesn't warn you if there are path parameters which don't get replaced. This could be improved in future by checking that there are no matches for `:` in the template after all replacements have been completed, except for the scheme at the start (if this is an absolute URL).